apollo-router/src/plugins/progressive_override/mod.rs

```diff
@@ -0,0 +1,246 @@
+    // don't "overprovide" any labels, since doing so can explode the number of
+    // cache entries per operation.
+    labels_per_operation_cache: Arc<DashMap<ParsedDocument, Vec<Arc<String>>>>,
+}
+
+type LabelsFromSchema = (HashMap<Arc<String>, Arc<f64>>, HashSet<Arc<String>>);
+
+fn collect_labels_from_schema(schema: &Schema) -> LabelsFromSchema {
+    let Some(directive_name) = spec::Schema::directive_name(
+        schema,
+        JOIN_SPEC_BASE_URL,
+        JOIN_SPEC_VERSION,
+        JOIN_FIELD_DIRECTIVE_NAME,
+    ) else {
+        tracing::error!(
+            "No join directive >=v0.4 found in the schema. No labels will be overridden."
+        );
+        return (HashMap::new(), HashSet::new());
+    };
+
+    let all_override_labels = schema
+        .types
+        .values()
+        .filter_map(|extended_type| {
+            if let ExtendedType::Object(object_type) = extended_type {
+                Some(object_type)
+            } else {
+                None
+            }
+        })
+        .flat_map(|object_type| &object_type.fields)
+        .filter_map(|(_, field)| {
+            let join_field_directives = field
+                .directives
+                .iter()
+                .filter(|d| d.name.as_str() == directive_name)
+                .collect::<Vec<_>>();
+            if !join_field_directives.is_empty() {
+                Some(join_field_directives)
+            } else {
+                None
+            }
+        })
+        .flatten()
+        .filter_map(|join_directive| {
+            if let Some(override_label_arg) =
+                join_directive.argument_by_name(OVERRIDE_LABEL_ARG_NAME)
+            {
+                override_label_arg
+                    .as_str()
+                    .map(|str| Arc::new(str.to_string()))
+            } else {
+                None
+            }
+        })
+        .collect::<HashSet<_>>();
+
+    let (percentages, other_labels): (HashSet<_>, HashSet<_>) = all_override_labels
+        .into_iter()
+        .partition(|label| label.starts_with("percent("));
+
+    let static_percentages = percentages
+        .into_iter()
+        .filter_map(|unparsed_label| {
+            unparsed_label
+                .strip_prefix("percent(")
+                .and_then(|unparsed_label| unparsed_label.strip_suffix(')'))
+                .and_then(|percent_as_string| percent_as_string.parse::<f64>().ok())
+                .map(|parsed_float| (Arc::new(unparsed_label.to_string()), Arc::new(parsed_float)))
+        })
+        .collect::<HashMap<_, _>>();
+
+    tracing::debug!("static_percentages: {:?}", &static_percentages);
+    (static_percentages, other_labels)
+}
+
+#[async_trait::async_trait]
+impl Plugin for ProgressiveOverridePlugin {
+    type Config = Config;
+
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        let schema = Schema::parse(&*init.supergraph_sdl, "schema.graphql").expect(
```

Comment:
not an issue with this PR, but we should make the parsed schema available to plugins, at the very least for internal plugins

Reply:
I took a stab at this, but happy to revert if it belongs in a different PR. Let me know what you think.
[5fd4a48](https://github.com/apollographql/router/pull/4521/commits/5fd4a488437b01f3eccdf9a870797985fda761f6)

Reply:
This was reverted based on a conversation we had out of thread